### PR TITLE
feat: add support for SHA-256 RSA PSS signatures 

### PR DESCRIPTION
**Pull request by madztheo**
# Description

Currently, the library only supports the PKCS#1 v1.5 scheme for RSA signatures. However, there is also another common scheme, namely RSA PSS. As it is more recent and considered more secure than PKCS#1 v1.5, adding support for it would be a great addition to the library. 

In the context of electronic passports, some countries are using RSA PSS, nearly exclusively with SHA-256. Thus, this PR aims to add support for the RSA PSS signature scheme using SHA-256 for the hash function and MGF1 for the mask generation function.

## Problem\*

Lack of support for RSA PSS signatures.

## Summary\*

This PR adds a new `verify_sha256_pss` function that can take in a signature generated with the PSS scheme and verify its validity against the provided message hash. It also requires to pass it the modulus size in bits as this value is required to decode the signature correctly. The exponent is assumed to be `65537`.

```rust
pub fn verify_sha256_pss(_: Self, instance: BNInstance, msg_hash: [u8; 32], sig: BN, key_size: u32) -> bool
```
A private function (not exposed to outside projects referencing the library) was also added to compute the MGF1 mask of a given seed, as it is required by the RSA PSS. 

```rust
fn mgf1_sha256<let SEED_LEN: u32, let MASK_LEN: u32>(seed: [u8; SEED_LEN]) -> [u8; MASK_LEN]
```
As this is not specific to RSA PSS, it might be worth exporting this function to a separate location and exposing it for other use cases.

Three test cases were added to make sure the verification logic works correctly.

```rust
#[test]
fn test_mgf1_sha256()
#[test]
fn test_verify_sha256_pss_2048()
#[test]
fn test_verify_sha256_pss_1964()
``` 
The first verifies that the mask generation function MGF1 works as expected. 
The second verifies a regular 2048-bit RSA signature. 
And the last verifies a more unusual 1964-bit RSA signature to test an edge case for key size not divisible by 8.

An argument was added to the `signature_gen` Rust package to generate an RSA PSS signature. As such, running the following command in `/signature_gen` will generate a PSS signature for the message "Hello World! This is Noir-RSA":

```bash
cargo run -- --msg "Hello World! This is Noir-RSA" --pss
```

## Additional Context

RSA PSS is more complex than PKCS#1 v1.5 and as such the gate count will be higher for verifying this type of signature. But the precise gate count has not been estimated yet.

# PR Checklist\*

- [x] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: lib/src/rsa.nr [hunk 1]
```diff
@@ -2,8 +2,74 @@ use dep::bignum::BigNum;
...
+
+fn reverse_array<let N: u32>(array: [u8; N]) -> [u8; N] {
+    let mut reversed = [0 as u8; N];
+    for i in 0..N {
+        reversed[i] = array[N - i - 1];
+    }
+    reversed
+}
+
+fn get_array_slice<let N: u32, let M: u32>(array: [u8; N], start: u32, end: u32) -> [u8; M] {
```
Comment (jfecher) [inline lib/src/rsa.nr:21]: Could use an `assert_eq(end - start, M);` in case users provide an incorrect length.

File: lib/src/rsa.nr [hunk 2]
```diff
@@ -48,6 +114,113 @@ fn compare_signature_sha256<let N: u32>(padded_sha256_hash: [u8; N], msg_hash: [
 }
 impl<BN, BNInstance, let NumBytes: u32> RSA<BN, BNInstance, NumBytes>  where BN: BigNumTrait, BNInstance: BigNumInstanceTrait<BN> {
     /**
+    * @brief Verify an RSA signature generated via the PSS signature scheme.
+    * @details `key_size` is the size of the RSA modulus in bits and is required to correctly decode the signature.
+    *
+    * @note We assume the public key exponent `e` is 65537
+    **/
+    pub fn verify_sha256_pss(
+        _: Self,
```
Comment (jfecher) [inline lib/src/rsa.nr:133]: Why not remove the Self parameter?

File: lib/src/rsa.nr [hunk 3]
```diff
@@ -48,6 +114,113 @@ fn compare_signature_sha256<let N: u32>(padded_sha256_hash: [u8; N], msg_hash: [
...
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
+        exponentiated = instance.mul(exponentiated, exponentiated);
```
Comment (jfecher) [inline lib/src/rsa.nr]:
```suggestion
        for _ in 0 .. 15 {
            exponentiated = instance.mul(exponentiated, exponentiated);
        }
```

File: lib/src/rsa.nr [hunk 4]
```diff
@@ -48,6 +114,113 @@ fn compare_signature_sha256<let N: u32>(padded_sha256_hash: [u8; N], msg_hash: [
...
+        // The length of the modulus in bytes
+        let key_len = (key_size + 7) / 8;
+        let h_len = 32;
+        let s_len = 32;
+
+        // Check if emLen < hLen + sLen + 2
+        assert(em_len >= h_len + s_len + 2);
+
+        // Check if eM ends with 0xBC
+        assert(em[em.len() - 1] == 0xBC);
```
Comment (jfecher) [inline lib/src/rsa.nr]:
```suggestion
        assert_eq(em[em.len() - 1], 0xBC);
```
Nitpick but in a few places you use `assert` where `assert_eq` would give better error messages since it'd show both values on error.

File: lib/src/rsa.nr [hunk 5]
```diff
@@ -48,6 +114,113 @@ fn compare_signature_sha256<let N: u32>(padded_sha256_hash: [u8; N], msg_hash: [
...
+        let masked_db: [u8; 223] = get_array_slice(em, offset, db_mask_len + offset);
+        let h = get_array_slice(em, db_mask_len + offset, em.len() - 1);
+
+        // Mask the leftmost bits
+        let bits_to_mask = 8 * em_len - em_bits;
+        let mask = 0xFF >> bits_to_mask as u8;
+        let first_byte = masked_db[offset];
+        // Make sure the 8 * em_len - em_bits leftmost bits are 0
+        // c.f. https://github.com/RustCrypto/RSA/blob/aeedb5adf5297892fcb9e11f7c0f6c0157005c58/src/algorithms/pss.rs#L205
+        assert((first_byte & !mask) == 0);
```
Comment (TomAFrench) [inline lib/src/rsa.nr]: This can be done more cheaply by just performing a division rather than bitwise operations.

File: lib/src/rsa.nr (hunk 1 above)
Comment (madztheo) [reply lib/src/rsa.nr:21]: Actually, this is not necessarily the case that `end - start` and `M` are equal, which is why I put this condition in the loop. If you take the following code for example:
```rust
let db_mask_len = em_len - h_len - 1;
let offset = key_len - em_len;
let masked_db: [u8; 223] = get_array_slice(em, offset, db_mask_len + offset);
```
`db_mask_len` will vary as there will be trailing zeros for key sizes below 2048 bits, 223 being the value `db_mask_len` will take for a 2048-bit public key. So because the array has to be constant-sized `end - start` will be less than or equal to M. I added an assert to check this inequality though. And I added some details in the comments to explain the reasoning.

Also, I ended up increasing the size of the array to 479 to account for key sizes up to 4096 bits, as with 223 it will not work for anything above 2048 bits. But this increased the number of constraints substantially (+30k). Which brings me to something that would truly help in this kind of situations, being able to do operations with generics and constant numbers for the size of arrays like so:
```rust
let masked_db: [u8; NumBytes - 32 - 1] = get_array_slice(em, offset, db_mask_len + offset);
```
This would make the code much more efficient, but it's not possible right now even if I don't see any reason as to why this shouldn't be computable at compile time.

File: lib/src/rsa.nr (hunk 2 above)
Comment (madztheo) [reply lib/src/rsa.nr:133]: As far as I know, this is required for methods of a Struct. At least the compiler won't let me do without it

File: lib/src/rsa.nr (hunk 4 above)
Comment (madztheo) [reply lib/src/rsa.nr]: Good point, I replaced all the `assert` checking an equality with `assert_eq`

File: lib/src/rsa.nr (hunk 3 above)
Comment (madztheo) [reply lib/src/rsa.nr]: Good point, it's much cleaner to read and seems to yield the same amount of constraints

File: lib/src/rsa.nr (hunk 5 above)
Comment (madztheo) [reply lib/src/rsa.nr]: Indeed, I switched to using powers of 2 now

File: lib/src/rsa.nr (hunk 1 above)
Comment (madztheo) [reply lib/src/rsa.nr:21]: Coming to think of it, the best trade-off would be to do like this in the meantime:
```rust
let masked_db: [u8; NumBytes] = get_array_slice(em, offset, db_mask_len + offset);
```
The array will always be 33 bytes too long, but that's better than having a fixed perfect size for 4096-bit RSA that will add unnecessary constraints for most people as the most common size remains 2048 bits. Plus it brings more flexibility as the max size is no longer hard-coded.

File: lib/src/rsa.nr [hunk 6]
```diff
@@ -48,6 +126,105 @@ fn compare_signature_sha256<let N: u32>(padded_sha256_hash: [u8; N], msg_hash: [
...
+        // But for now we can't so we'll just use NumBytes and have 33 trailing 0s
+        let masked_db: [u8; NumBytes] = get_array_slice(em, offset, db_mask_len + offset);
+        let h = get_array_slice(em, db_mask_len + offset, em.len() - 1);
+
+        // Make sure the 8 * em_len - em_bits leftmost bits are 0
+        // c.f. https://github.com/RustCrypto/RSA/blob/aeedb5adf5297892fcb9e11f7c0f6c0157005c58/src/algorithms/pss.rs#L205
+        let bits_to_mask = 8 * em_len - em_bits;
+        let mask_value = pow(2, bits_to_mask as u32);
+        let max_allowed_value = 255 / mask_value;
+        assert(masked_db[0] as u32 <= max_allowed_value);
```
Comment (TomAFrench) [inline lib/src/rsa.nr]:
```suggestion
        let bits_to_mask = 8 * em_len - em_bits;
        let mask_value = pow(2, bits_to_mask as u32);
        assert_eq(masked_db[0] as u32 / mask_value, 0);
```

File: lib/src/rsa.nr (hunk 1 above)
Comment (jfecher) [reply lib/src/rsa.nr:21]: > This would make the code much more efficient, but it's not possible right now even if I don't see any reason as to why this shouldn't be computable at compile time.

Assuming `NumBytes` is the same generic as is on the impl, doing arithmetic on it is banned by most languages including Rust because it makes type checking more difficult/impossible in some cases (simplifying here). We actually recently merged this feature into Noir however, so this code is allowed by default now. There's some documentation for it here: https://noir-lang.org/docs/dev/noir/concepts/generics#arithmetic-generics. If you're using a stable release you may have to wait some time for a new release to include this change.

File: lib/src/rsa.nr (hunk 2 above)
Comment (jfecher) [reply lib/src/rsa.nr:133]: You should be able to remove it, but the method may need to be moved to a separate impl since the generics on the impl should no longer apply either. Since there'd be no self type you'd also have to call it via the fully qualified syntax.

File: lib/src/rsa.nr (hunk 1 above)
Comment (madztheo) [reply lib/src/rsa.nr:21]: Alright, good to know that's supported now. I'm using Noir 0.32 on my machine at the moment. I assume you're referring to Noir 0.33. I've tried using Noir 0.33 before, but a few breaking changes broke the big int library, so I reverted back to Noir 0.32. But yes, it looks like the linter is not bugging me anymore when I try the code I mentioned with nargo 0.33. I'll try to fix the issues with the big int library (at least locally) to get it to compile and run the tests and I'll commit the changes.

File: lib/src/rsa.nr (hunk 6 above)
Comment (madztheo) [reply lib/src/rsa.nr]: This is actually not the desired behavior. What we want to check is that all the `bits_to_mask` leftmost bits are all equal to 0. Here, it's checking that the `8 - bits_to_mask` leftmost bits are zero. But following your logic, I'll replace my code with this:
```rust
let bits_to_mask = 8 - (8 * em_len - em_bits);
let mask_value = pow(2, bits_to_mask as u32);
assert_eq(masked_db[0] as u32 / mask_value, 0);
 ```

File: lib/src/rsa.nr (hunk 1 above)
Comment (madztheo) [reply lib/src/rsa.nr:21]: Just pushed it. You can have a look at the necessary fixes for the big int library here https://github.com/madztheo/noir-bignum/commit/d0c211d605ce12af3b2aab51317e9681189c3f2a

This means that the lib only works with Noir 0.33 and above though. But at the same time, the library doesn't work in its current state with Noir 0.33 because of the issues with the big int library due to the `to_le_bytes` function not taking any argument anymore.

File: lib/src/rsa.nr (hunk 2 above)
Comment (madztheo) [reply lib/src/rsa.nr:133]: These changes have been applied with which commit of the Noir repo exactly?

Comment (jfecher) [reply lib/src/rsa.nr:133]: Noir has never required `self` to be in an impl method, you just can't call the method as `foo.bar()` without a self parameter. You'd need to call it as `Foo::bar(foo)`, and potentially move it to another impl so that the generics aren't unused.

The only change is the last point where we issue an error now if a function has unused generics, even from the outer impl. This change was done in https://github.com/noir-lang/noir/pull/5840

Comment (madztheo) [reply lib/src/rsa.nr:133]: I don't really see how to use it in this case though since the generics are at the level of the impl and not the function itself. So if I try something like this:
```rust
RSA::<BN2048, BNInst2048, 256>::verify_sha256_pss(BNInstance, sha256_hash, signature, 2048);
```
or
```rust
type RSA2048 = RSA<BN2048, BNInst2048, 256>;
RSA2048::verify_sha256_pss(BNInstance, sha256_hash, signature, 2048)
```
None of the two works. Also, I believe the point of how this lib has been developed is to abstract away all the generics so at the end of the day you can just use something like RSA2048 with everything already predefined in the library. For example, the original PKCS#1 signature verification:
```rust
let rsa: RSA2048 = RSA {};
assert(rsa.verify_sha256_pkcs1v15(BNInstance, sha256_hash, signature));
```

Comment (jfecher) [reply lib/src/rsa.nr:133]: > I don't really see how to use it in this case though since the generics are at the level of the impl and not the function itself. So if I try something like this:

Noir doesn't have the ability to specify generics on an impl type yet unfortunately so `Foo::<generics>::method(...)` won't work in general. You'd have to move the generics on to the function itself in that case and do `verify_sha256_pss::<BN2048, BNInst2048, 256>(... args ...)`.

> Also, I believe the point of how this lib has been developed is to...

That's another valid approach if that's how we want to specify the generics - sorry for the detour.
If we removed the object we'd need a different version of this function (and others) specialized to the generic types instead which would be M functions times N configurations instead of just M functions plus N struct configurations as in the existing scheme.

**Comment by TomAFrench** [comment]
This has merge conflicts with master.

**Comment by madztheo** [comment]
> This has merge conflicts with master.

I've resolved the conflicts. It should be good to go now

**Comment by kashbrti** [comment]
@madztheo seems like this PR is from before the bignum refactor. do you guys have a version with the new bignum or should I make the changes accordingly?

**Comment by madztheo** [comment]
@kashbrti We do have a version working with the latest bignum on our GitHub org https://github.com/zkpassport/noir_rsa
